### PR TITLE
[node] Add generics to 'EventEmitter'

### PR DESCRIPTION
**Pull request by malthe**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/25302.
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by malthe** [comment]
This fixes issue #25302.

**Comment by Semigradsky** [comment]
Nice try

**Comment by malthe** [comment]
I did a quick analysis of the errors in packages that depend on `EventEmitter` and extending the type spec in an incompatible way (not taking generics into account):
- amqp
- better-queue
- node-red
- http-proxy
- express-serve-static-core
- sse
- bonjour
- steam-client
- steam-user
- globaloffensive
- steam-tradeoffer-manager
- chromecast-api
- clarinet
- sap__approuter
- umzug
- degit
- duplex-to
- file-manager
- formidable
- gc-stats
- gulp-javascript-obfuscator
- ignore-walk
- imap
- mail-notifier
- node-imap
- jake
- kahoot.js-updated
- memcached
- midi
- mock-knex
- nedb
- nestdb
- node-localstorage
- node-resque
- node-ssdp
- node-vagrant
- nodemailer
- opossum
- orchestrator
- parse
- pg
- pg
- newman
- proggy
- rdf-store-fs
- rdf-transform-triple-to-quad
- readable-stream
- rdf-dataset-ext
- twitter
- sematext__gc-stats
- simply-blackjack
- steam
- streamx
- stylus
- xml-flow
- amqp
- better-queue
- node-red
- http-proxy
- express-serve-static-core
- sse
- bonjour
- steam-client
- steam-user
- globaloffensive
- steam-tradeoffer-manager
- chromecast-api
- clarinet
- sap__approuter
- umzug
- degit
- duplex-to
- file-manager
- formidable
- gc-stats
- gulp-javascript-obfuscator
- ignore-walk
- imap
- mail-notifier
- node-imap
- jake
- kahoot.js-updated
- loopbench
- memcached
- midi
- mock-knex
- nedb
- nestdb
- node-localstorage
- node-resque
- node-ssdp
- node-vagrant
- nodemailer
- opossum
- orchestrator
- parse
- pg
- pg
- newman
- proggy
- rdf-store-fs
- rdf-transform-triple-to-quad
- readable-stream
- rdf-dataset-ext
- twitter
- sematext__gc-stats
- simply-blackjack
- steam
- streamx
- stylus
- xml-flow

In total 110 packages, each would require some attention for a clean run.

**Comment by malthe** [comment]
I have worked through the dependencies up until "globaloffensive" and have solidified the `EventEmitter` generics to accomodate the needs encountered thus far.

Specifically, there seems to be a need for separating the `on(...)` family from `emit(...)` – that is, sometimes packages want to emit a subset of the event that can be subscribed to.

This seems reasonable and means there are now two generic variables, `T` and `E` (where the latter defaults to the former which in turn is optional.).

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/node/events.d.ts
1. types/node/test/events_generic.ts
1. types/node/ts4.8/test/events_generic.ts
1. types/node/ts4.8/events.d.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 430b6d95c23e1e2398deaf49eb9a5892a6f3019d
</p>

**Review by mcollina — APPROVED** [review]
LGTM, clever trick.

**Review by yoursunny — APPROVED** [review]
I wanted to do this long ago but didn't have the stomach to deal with some many dependants. I'm glad this is happening now.

**Comment by malthe** [comment]
I have added an example/test which demonstrates how to type a function that "proxies" a call into `EventEmitter`.

With TypeScript 4.8+ this can be written as:
```typescript
// eslint-disable-next-line @definitelytyped/no-unnecessary-generics
function on<K>(...args: Parameters<typeof emitter.on<K>>): void {
    emitter.on(...args);
}
```
Meanwhile on TypeScript 4.6, we're not able to use `...args` but instead have to provide concrete types for `eventName` and `listener`.

We don't currently export the `Listener` helper type; a package will currently have to define its own.

Here's an example of how a class might encapsulate an `EventEmitter` – assuming `Events` is an already defined event map.
```typescript
type Listener<K> = (
    K extends keyof Events ? (
            Events[K] extends unknown[] ? (...args: Events[K]) => void : Function
        )
        : Function
);

class MyClass {
    private readonly events = new EventEmitter<Events>();

    on<K>(...args: Parameters<typeof this.events.on<K>>): void {
        this.events.on(...args);
    }
```
Defining `on` this way obviates the need for specifying the arguably somewhat complex types.

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/node/test/events_generic.ts
1. types/node/ts4.8/test/events_generic.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 8d69148e8c4fb531b203b541afdf816db8eb6a53
</p>

File: types/node/test/events_generic.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    result = emitter.on(event1, listener1);
+    result = emitter.once(event1, listener1);
+    result = emitter.prependListener(event1, listener1);
+    result = emitter.prependOnceListener(event1, listener1);
+    result = emitter.removeListener(event1, listener1);
+    result = emitter.off(event1, listener1);
+    result = emitter.removeAllListeners();
+    result = emitter.removeAllListeners(event1);
+    result = emitter.setMaxListeners(42);
+
```
Comment (yoursunny) [inline types/node/test/events_generic.ts:36]: Consider testing with less strict listener types, such as:

```ts
emitter.addListener(event1, () => undefined);
emitter.addListener(event1, (a: string) => undefined);
emitter.addListener(event1, (a: string, b?: number) => undefined);
emitter.addListener(event1, (a: string, b: number | boolean): number => 1);
```

**Review by yoursunny — APPROVED** [review]

File: types/node/test/events_generic.ts (hunk 1 above)
Comment (malthe) [reply types/node/test/events_generic.ts:36]: Added (verbatim, plus one more that's intentionally erroneous) in 50dca4c91e05506dc215e9209691d144119bf95f.

**Review by yoursunny — APPROVED** [review]

**Comment by malthe** [comment]
@yoursunny tests currently failing for other reasons.

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/node/events.d.ts
1. types/node/ts4.8/events.d.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against e9bd57c3529b2a8aa90900cb63befa6655eb45f2
</p>

**Comment by malthe** [comment]
@Semigradsky mind taking another look at this?

What's arguably good is that tests are green, the test suite has been augmented with quite a few more tests to cover additional cases. Now, one can disagree with the implementation, it's perhaps not the most elegant, but my attempts to make it more clean/straightforward have failed because of the many expectations from existing code.

**Review by yoursunny — APPROVED** [review]

**Comment by jakebailey** [comment]
I'm going to close/reopen this just because I'm sure the CI results are stale.

My main concern is that this is not updating the types for the other versions; is that intentional? Is this exclusive to Node v20? It feels like it shouldn't be.

**Comment by malthe** [comment]
@jakebailey it's not intentionally limited to v20; I didn't know there was other supported versions in-tree.

**Comment by jakebailey** [comment]
Yeah; v16 and v18s types are still supported and should be updated if possible.

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/node/v18/events.d.ts
1. types/node/v16/events.d.ts
1. types/node/v18/ts4.8/events.d.ts
1. types/node/v16/ts4.8/events.d.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against bc3a6b97a63a17ace1b8b416653be0489d0c8280
</p>

**Comment by malthe** [comment]
@jakebailey definitely agree that the complexity in the implementation is considerable (and somehow not particularly elegant), but I think the upshot is that the tests demonstrate that you can use this new generic typing in a simple, straight-forward way. I think we can iterate further on the implementation post-merge.

**Review by jakebailey — APPROVED** [review]
Okay, then I guess we'll give it a try.

**Comment by malthe** [comment]
Ready to merge

**Comment by ckohen** [comment]
Is there a reason generics were not added to the static methods? It is totally possible to detect the event map even in the static methods and provide the same safety there. I would be happy to do so but wanted to confirm there wasn't a reason it wasn't done in the first place (couldn't see one in reading through the PR).

**Comment by malthe** [comment]
@ckohen no reason as far as I am aware.

**Pull request merged**